pkg/common-controller/groupsnapshot_controller_helper.go

```diff
@@ -935,3 +950,283 @@ func (ctrl *csiSnapshotCommonController) needsUpdateGroupSnapshotStatus(groupSna
+	if err != nil {
+		return err
+	}
+	// check whether the group snapshot content points back to the passed in group
+	// snapshot, note that binding should always be bi-directional to trigger the
+	// deletion on group snapshot content or adding any annotation to the group
+	// snapshot content
+	var deleteGroupSnapshotContent bool
+	if groupSnapshotContent != nil && utils.IsVolumeGroupSnapshotRefSet(groupSnapshot, groupSnapshotContent) {
+		// group snapshot content points back to group snapshot, whether or not
+		// to delete a group snapshot content now depends on the deletion policy
+		// of it.
+		deleteGroupSnapshotContent = (groupSnapshotContent.Spec.DeletionPolicy == crdv1.VolumeSnapshotContentDelete)
+	} else {
+		// the group snapshot content is nil or points to a different group snapshot, reset group snapshot content to nil
+		// such that there is no operation done on the found group snapshot content in
+		// checkandRemoveSnapshotFinalizersAndCheckandDeleteContent
+		groupSnapshotContent = nil
+	}
+
+	klog.V(5).Infof("processGroupSnapshotWithDeletionTimestamp[%s]: delete group snapshot content and remove finalizer from group snapshot if needed", utils.GroupSnapshotKey(groupSnapshot))
+
+	return ctrl.checkandRemoveGroupSnapshotFinalizersAndCheckandDeleteGroupSnapshotContent(groupSnapshot, groupSnapshotContent, deleteGroupSnapshotContent)
+}
+
+// checkandRemoveGroupSnapshotFinalizersAndCheckandDeleteGroupSnapshotContent deletes
+// the group snapshot content and removes group snapshot finalizers if needed
+func (ctrl *csiSnapshotCommonController) checkandRemoveGroupSnapshotFinalizersAndCheckandDeleteGroupSnapshotContent(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot, groupSnapshotContent *crdv1alpha1.VolumeGroupSnapshotContent, deleteGroupSnapshotContent bool) error {
+	klog.V(5).Infof("checkandRemoveGroupSnapshotFinalizersAndCheckandDeleteGroupSnapshotContent VolumeGroupSnapshot[%s]: %s", utils.GroupSnapshotKey(groupSnapshot), utils.GetGroupSnapshotStatusForLogging(groupSnapshot))
+
+	if !utils.IsGroupSnapshotDeletionCandidate(groupSnapshot) {
+		return nil
+	}
+
+	// TODO: In this needed?
+	// check if an individual snapshot belonging to the group snapshot is being
+	// used for restore a PVC
+	// If yes, do nothing and wait until PVC restoration finishes
+	for _, snapshotRef := range groupSnapshot.Status.VolumeSnapshotRefList {
+		snapshot, err := ctrl.snapshotLister.VolumeSnapshots(snapshotRef.Namespace).Get(snapshotRef.Name)
+		if err != nil {
+			return err
+		}
+		if ctrl.isVolumeBeingCreatedFromSnapshot(snapshot) {
+			klog.V(4).Infof("checkandRemoveGroupSnapshotFinalizersAndCheckandDeleteGroupSnapshotContent[%s]: snapshot [%s] belonging to volume group snapshot is being used to restore a PVC", utils.GroupSnapshotKey(groupSnapshot), utils.SnapshotKey(snapshot))
+			ctrl.eventRecorder.Event(snapshot, v1.EventTypeWarning, "SnapshotDeletePending", "Snapshot is being used to restore a PVC")
```

Comment:
Can you print out the group snapshot name as well?

Reply:
In the event or in the log?

Reply:
I think the event should added to the groupSnapshot, not the snapshot, because we are trying to delete a groupSnapshot here.  You can print out the snapshot name in the message.

Reply:
Done, I changed the log message and the event here